Feature/add react header

adamledwards opened the pull request:
# React Header
1. Added the useMODUKComponent and started to type the components we will use from it. 
2. Added SVGR and config to transpile `react-merge-ref`. TS-node is still present with transpileOnly turned on.
3. Added a hydrationError to the UI so that the E2E will fail
4. Added the Header component and some more testes 
5. React project is tested against the current tests we have.

adamledwards commented on .eleventy.js:
```diff
@@ -1,8 +1,35 @@
+    },
+  }],
+})
+
```
When with svgr because I can isolate the SVG file, Otherwise if I went with babel inline, I would need to do the transformation on the TSX via babel.

adamledwards replied:
switched this to svg inline

adamledwards commented on examples-site/react/example-client.tsx:
```diff
@@ -2,6 +2,24 @@ import type { ComponentType } from 'react'
 import { StrictMode } from 'react'
 import { hydrateRoot } from 'react-dom/client'
 
+const errorDom = (
+  title: string,
+  info?: string,
+) => (
+  `<div class="govuk-error-summary" data-module="govuk-error-summary">
+  <div role="alert">
+    <h2 class="govuk-error-summary__title">${title}</h2>
+    ${
+    info && `<div class="govuk-error-summary__body">
+      <ul class="govuk-list govuk-error-summary__list">
+        <li>${info}</li>
+      </ul>
+    </div>`
+  }
+  </div>
+</div>`
+)
+
```
I wanted to test the hydrationRoot function that works in all our examples 
 
![image](https://github.com/defencedigital/moduk-frontend/assets/5575331/547ab698-9a66-4114-8d7f-3a640cf6c669)
So I added error for our E2E test will fail if there is a mismatch

adamledwards commented on examples-site/react/example-client.tsx:
```diff
@@ -13,7 +31,7 @@ async function getComponent(): Promise<ComponentType> {
 
   return (await import(
-    /* webpackInclude: /\/\w+\/__examples__\/(\w|-)+\.tsx$/ */
+    /* webpackInclude: /\/(\w|-)+\/__examples__\/(\w|-)+\.tsx$/ */
```
This was not picking up the back link example `/back-link/__examples__/default.tsx`

adamledwards commented on playwright.config.ts:
```diff
@@ -1,7 +1,40 @@
+    },
+  ])
+}
+
```
This was fun. 
I added react projects based on the nunjuck project, 
This function ass prefixes react to the project name e.g `react-firefox` and will only test for the  React components that have been implemented. 

I've also added inverted greps for `@react` and `@nunjucks` see the `visual-regression.spec.ts` and `accesibility.spec.ts` for their usage

adamledwards commented on src/lib/utils.ts:
```diff
@@ -48,6 +49,45 @@ export function addMODUKFilters(nunjucksEnv: Environment) {
+ * Based from https://github.com/11ty/eleventy/blob/v2.0.1/src/Engines/Nunjucks.js#L166-L228
+ * as https://mozilla.github.io/nunjucks/api.html#custom-tags
+ */
+export function addSvgExtension(nunjucksEnv: Environment) {
+  class SvgLoader {
+    tags = ['svg']
+
+    /* eslint-disable @typescript-eslint/no-explicit-any */
+    parse(parser: any, nodes: any) {
+      const tok = parser.nextToken()
+
+      const args = parser.parseSignature(true, true)
+
+      // Nunjucks bug with non-paired custom tags bug still exists even
+      // though this issue is closed. Works fine for paired.
+      // https://github.com/mozilla/nunjucks/issues/158
+      if (args.children.length === 0) {
+        args.addChild(new nodes.Literal(0, 0, ''))
+      }
+
+      parser.advanceAfterBlockEnd(tok.value)
+
+      return new nodes.CallExtension(this, 'run', args)
+    }
+    /* eslint-enable @typescript-eslint/no-explicit-any */
+
+    /* eslint-disable class-methods-use-this */
+    run(_context: unknown, fileName: string) {
+      const svgPath = `${join(__dirname, '../assets/svg', basename(fileName, '.svg'))}.svg`
+      return new runtime.SafeString(readFileSync(svgPath, 'utf8'))
+    }
+    /* eslint-enable class-methods-use-this */
+  }
+  nunjucksEnv.addExtension('MODUKSvgLoader', new SvgLoader())
+}
+
```
reimplemented the SVG shortcode in the moduk environment, before it was in 11ty but then realised that not going to work for portability. 
https://github.com/11ty/eleventy/blob/v2.0.1/src/Engines/Nunjucks.js#L166-L228
https://mozilla.github.io/nunjucks/api.html#custom-tags

adamledwards replied:
Upon reflection I should update the tag name to something less generic like MODUKSvg

adamledwards commented on src/react/hooks/useMODUKComponent/useMODUKComponent.ts:
```diff
@@ -0,0 +1,38 @@
+import { useEffect, useRef } from 'react'
+
+type Component = {
+  type: 'Header'
+  element: HTMLElement
+}
+
+type ComponentNameUnion<T extends { type: string; element: HTMLElement }> = T['type']
+
+type ComponentName = ComponentNameUnion<Component>
+
+type FindElementType<C extends ComponentName> = Extract<Component, { type: C }>['element']
+
```
This ensures that the correct ref type is returned for the element.

For example to add a Button component you would have 
```ts
// useMODUKComponent.ts
type Component = {
  type: 'Header'
  element: HTMLElement
} | 
{
  type: 'Button'
  element: HTMLButtonElement
}
```

usage
```tsx

 () => {
  const { ref: buttonRef } =useMODUKComponent('Button')
 // typeof buttonRef === HTMLButtonElement
 
   const { ref: headerRef } =useMODUKComponent('Header')
 // typeof headerRef === HTMLElement
}
```

jpveooys reviewed (commented):
Really nice work!

Some other things to think about (probably mostly for a separate PR as there's enough going on here already :smile: ):

- build types for the React components
- fix/inline SVGs in the built React components
- update `exports` in `package.json`
- exclude React tests/examples from `dist` (e.g. pass `--ignore 'src/react/**/__examples__' --ignore 'src/react/**/__tests__'` to `babel`)

Also would be good to add something that does a comparison of the generated markup of the React and Nunjucks examples, but also think it'll be difficult even with a minifier due to whitespace differences.

Also FYI when in another TypeScript project how the import is done depends on `moduleResolution` in `tsconfig.json`. If it's `"moduleResolution": "node"` then you have to import from `dist/`: `import { Header } from '@moduk/frontend/dist/react'`.

If `moduleResolution` is set to `bundler` (or `node16` or `nodenext`) then `import { Header } from '@moduk/frontend/react'` works.

jpveooys commented on playwright.config.ts:
```diff
@@ -1,7 +1,40 @@
+  ])
+}
+
 expect.extend({
```
Might be good to move the `expect.extend` stuff higher up so the project stuff is together?

adamledwards commented on src/react/hooks/useMODUKComponent/useMODUKComponent.ts:
```diff
@@ -0,0 +1,38 @@
+type FindElementType<C extends ComponentName> = Extract<Component, { type: C }>['element']
+
+const initialiseComponent = async (componentName: ComponentName, el: HTMLElement) => {
+  const client = await import('@moduk/frontend/client')
```
I can try to see if we can import the esm module, (In a new PR 😅 )

adamledwards commented:
@jpveooys generating the types to not include `@moduk/fontend/client` is proving to be difficult, do you| mind having a look 
I'm looking to generate a structure like this
```
dist
├── react
│   └── types
│           └── index.d.ts
``` 

what I'm getting is this 

```
dist
├── react
│   └── types
│           └── react
│                 └── index.d.ts
│           └── client
│                  └── index.d.ts
```

jpveooys commented:
@adamledwards Seems like it's managing to pick up the import of `@moduk/frontend/client` and then creating types for `client/index.ts` as well. So you end up with `react` and `client` subdirectories. Not sure there's a way to overcome that (`exclude` only affects what `include` matches).

Having said that, it might be good to put the types alongside the generated `.js` files anyway (e.g. `dist/react/index.js`, `dist/react/index.d.ts`). That way they'll get picked up even if importing directly from the `dist` directory.

adamledwards commented:
> That way they should get picked up even if importing directly from the `dist` directory (if TS has `"moduleResolution" : "node"`).
According to this
https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#ecmascript-module-support-in-node-js
`"moduleResolution" : "node"` won't work with our library because we are using exports in our package.json 

I've added it alongside the esm build, not sure if it should be the cjs build or if it matters. I've removed the types from `@moduk/frontend/client` and typed `govuk-frontend` instead.

jpveooys commented:
> According to this
https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#ecmascript-module-support-in-node-js
"moduleResolution" : "node" won't work with our library because we are using exports in our package.json

It won't understand `exports`, nope, but you can still import from `dist`. So we'll probably use to import from `dist` in our examples to make things work universally (probably the guidance site would have to change the code snippets to do that as I don't think we can do it here cleanly).

jpveooys reviewed (commented):
Nice work and updates!

We still need to sort the `js-enabled` problem, right? (so that it just works for users – the example site still has the inline script in.)

I wonder if we should leave out `npm run build:client:react` from the `build` script for now so we can get this merged and create another ticket to do some more testing with different frameworks and mop up any problems remaining.

jpveooys commented:
@adamledwards FYI there's a test failing.

jpveooys commented:
Did you see this comment @adamledwards https://github.com/defencedigital/moduk-frontend/pull/275#discussion_r1238534989? Only one left I think

jpveooys reviewed (approved):
Nice!
